torchrl/modules/models/model_based.py

```diff
@@ -0,0 +1,292 @@
+
+
+    """
+
+    def __init__(self, rssm_prior: TensorDictModule, rssm_posterior: TensorDictModule):
+        super().__init__()
+        _module = TensorDictSequential(rssm_prior, rssm_posterior)
+        self.in_keys = _module.in_keys
+        self.out_keys = _module.out_keys
+        self.rssm_prior = rssm_prior
+        self.rssm_posterior = rssm_posterior
+
+    def forward(self, tensordict):
+        """Runs a rollout of simulated transitions in the latent space given a sequence of actions and environment observations.
+
+        The rollout requires a belief and posterior state primer.
+
+        At each step, two probability distributions are built and sampled from:
+        - A prior distribution p(s_{t+1} | s_t, a_t, b_t) where b_t is a
+            deterministic transform of the form b_t(s_{t-1}, a_{t-1}). The
+            previous state s_t is sampled according to the posterior
+            distribution (see below), creating a chain of posterior-to-priors
+            that accumulates evidence to compute a prior distribution over
+            the current event distribution:
+            p(s_{t+1} s_t | o_t, a_t, s_{t-1}, a_{t-1}) = p(s_{t+1} | s_t, a_t, b_t) q(s_t | b_t, o_t)
+
+        - A posterior distribution of the form q(s_{t+1} | b_{t+1}, o_{t+1})
+            which amends to q(s_{t+1} | s_t, a_t, o_{t+1})
+
+        """
+        tensordict_out = []
+        *batch, time_steps = tensordict.shape
+        _tensordict = tensordict[..., 0]
+
+        update_values = tensordict.exclude(*self.out_keys)
+        for t in range(time_steps):
+            # samples according to p(s_{t+1} | s_t, a_t, b_t)
+            # ["state", "belief", "action"] -> ["next_prior_mean", "next_prior_std", "_", "next_belief"]
+            self.rssm_prior(_tensordict)
+
+            # samples according to p(s_{t+1} | s_t, a_t, o_{t+1}) = p(s_t | b_t, o_t)
+            # ["next_belief", "next_encoded_latents"] -> ["next_posterior_mean", "next_posterior_std", "next_state"]
+            self.rssm_posterior(_tensordict)
+
+            tensordict_out.append(_tensordict)
+            if t < time_steps - 1:
+                _tensordict = step_mdp(
+                    _tensordict.select(*self.out_keys), keep_other=False
+                )
+                _tensordict = update_values[..., t + 1].update(_tensordict)
+
+        return torch.stack(tensordict_out, tensordict.ndimension() - 1).contiguous()
+
+
+class RSSMPrior(nn.Module):
+    """The prior network of the RSSM.
+
+    This network takes as input the previous state and belief and the current action.
+    It outputs the next prior state and belief, as well as the parameters of the prior state distribution.
+    State is by construction stochastic and belief is deterministic. In the "Dream to control" paper, these are called "deterministic state " and "stochastic state", respectively.
+    We distinguish states sampled according prior and posterior distribution for clarity.
```

Comment:
Not true anymore (+ not necessarily clearer since we sometimes have to rename them)